docs/src/abigen/the-abigen-macro.md

```diff
@@ -0,0 +1,106 @@
+```text
+ProgramType(name="MyProgramType", abi="my_program-abi.json")...
+```
+where:
+
+`ProgramType` is one of: `Contract`, `Script` or `Predicate`,
+
+`name` is the name that will be given to the generated bindings,
+
+`abi` is either a path to the json abi file or its actual contents.
+
+---
+So, an `abigen!` which generates bindings for two contracts and one script looks like this:
+```rust,ignore
+{{#include ../../../examples/abigen/src/lib.rs:multiple_abigen_program_types}}
+```
+
+## How does the generated code look?
+
+A rough overview:
+```rust,ignore
+pub mod abigen_bindings {
+    pub mod contract_a_mod {
+        struct SomeCustomStruct{/*...*/};
+        // other custom types used in the contract
+        
+        struct ContractA {/*...*/};
+        impl ContractA {/*...*/};
+        // ...
+    }
+    pub mod contract_b_mod {
+        // ...
+    }
+    pub mod my_script_mod {
+        // ...
+    }
+    pub mod my_predicate_mod{
+        // ...
+    }
+    pub mod shared_types{
+        // ...
+    }
+}
+
+pub use contract_a_mod::{/*..*/};
+pub use contract_b_mod::{/*..*/};
+pub use my_predicate_mod::{/*..*/};
+pub use shared_types::{/*..*/};
+```
+
+Each `ProgramType` gets its own `mod` based on the `name` given in the `abigen!`. Inside the respective mods the custom
+types used by that program are generated, as well as the bindings through which the actual calls can be made.
+
+There is one extra `mod` called `shared_types` which is generated if `abigen!` detects that the given programs share
+types. Instead of each `mod` regenerating the type for itself, the type is lifted out into the `shared_types` module and
+generated only once and then shared between all programs that use it.
+
+A type is deemed shared if its name and definition match up. This can happen either because you've used the same library
+(a custom one or a type from the stdlib) or because you've happened to define the exact same type.
+
+Finally `pub use` statements are inserted, so you don't have to fully qualify the generated types. To avoid conflict
+only types which have unique names will get a `pub use` statement. If you find rustc can't find your type it might just
+be that there is another generated type with the same name. To fix the issue just qualify the path by doing
+`abigen_bindings::whatever_contract_mod::TheType`.
+
+> **Note:**
+> It is **highly** encouraged that you generate all your bindings in one `abigen!` call. Doing it in this manner will
+allow type sharing and avoid  name collisions you'd normally get when calling `abigen!` multiple times inside the same
+namespace. If you choose to proceed otherwise, then keep in mind the generated code overview presented above and
+separate the `abigen!` calls into different modules appropriately to resolve the collision.
+
+## Using the bindings
+Let's look at a contract with two methods: `initialize_counter(arg: u64) -> u64` and `increment_counter(arg: u64) -> u64`, with the following JSON ABI:
```

Comment:
Same here: no break lines in a paragraph + wording improvements.

```suggestion
> It is **highly** encouraged that you generate all your bindings in one `abigen!` call. Doing it in this manner will allow type sharing and avoid name collisions you'd normally get when calling `abigen!` multiple times inside the same namespace. If you choose to proceed otherwise, keep in mind the generated code overview presented above and appropriately separate the `abigen!` calls into different modules to resolve the collision.

## Using the bindings
Let's look at a contract with two methods: `initialize_counter(arg: u64) -> u64` and `increment_counter(arg: u64) -> u64`, with the following JSON ABI:
```